### DISCUSSION TITLE
Where can I find documentation of textlayout module of skia?

### DISCUSSION THREAD

**Discussion by Fancyflame** [rust-skia/rust-skia#792] [Q&A] [answered]
I'm using the `textlayout` feature of `skia-safe`, but there is no documentation found in the rust crate. Then I turned to [api documentation at official site](https://api.skia.org/), but only those about fundamental apis of skia found. All I found about `textlayout` is just some preview and summary. Where can I find detailed documentation of API? Thanks

**Comment by pragmatrix** [comment, marked as answer]
You can take a look at the Skia C header files at https://github.com/google/skia/tree/main/modules/skparagraph/include, although documentation looks thin in there, too. And also post a request to the [Skia issue tracker](https://skia.org/docs/user/issue-tracker/), asking why the documentation of the skparagraph module is missing from the official API documentation site. Meanwhile, I will try my best to convert the existing comments. Progress gets posted to #446.

**Reply by pragmatrix** [reply]
See also #793

**Reply by Fancyflame** [reply]
Those comments are barely enough to use, thanks a lot.

**Reply by pragmatrix** [reply]
I've build a rudimentary text box editor and renderer using Skia Safe. While I haven't made the project itself open source, here are the interesting files: 
- Renderer: https://gist.github.com/pragmatrix/9f37662327cd41443b44e0f1b587f592 
- Editor: https://gist.github.com/pragmatrix/0c7b8be33d9f54982fd47789fffd543a

May be this helps a little. Also note that this is based on an older version of skia-safe, so it might need some adjustments.

**Reply by Fancyflame** [reply]
That's a big help! That's exactly what I am trying to achieve in my GUI framework. I'll read it later. 😄
Btw, this project seems interesting, are you going to make it closed source? Maybe it is a pity.
